### ISSUE TITLE
RFC: `units`: wrap any array API standard compatible array

### ISSUE THREAD

**Issue by lucascolley** [astropy/astropy#17588]
*I just posted a very similar issue over at `pint`, but posting here too!*

---

[The array API standard](https://data-apis.org/array-api/latest/index.html) is being utilised in libraries like SciPy and scikit-learn to add support for array libraries like CuPy, PyTorch, and JAX. It seems feasible that `astropy.units` could wrap any library which is compatible with the standard.

The model for how this can be done is demonstrated in https://github.com/mdhaber/marray. Marray adds masked array support to standard-compatible arrays. Similar to `Quantity`, an `MArray` is some `data` together with a `mask`:

```
MArray(
    Array([1, 2], dtype=array_api_strict.int64),
    Array([False,  True], dtype=array_api_strict.bool)
)
```

Marray exposes one function, `get_namespace`, which takes as input a standard-compatible namespace, and returns a new standard-compatible namespace with the additional capabilities of creating and correctly processing masked arrays. Under the hood, this namespace wraps each function to handle anything specific to masks, and otherwise dispatches to the underlying array namespace. So here, `astropy.units.get_namespace(torch)` would return a namespace for astropy-unit-wrapped PyTorch tensors.

`MArray`s also expose this namespace via the [`__array_namespace__`](https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__array_namespace__.html#array-namespace) method, which allows libraries to operate on them like any other standard array:

```python
# get the namespace for input astropy-units(torch) array x
xp = x.__array_namespace__()
y = xp.exp(x) # returns another astropy-units(torch) array
z = xp.asarray([1, 2, 3]) # also an astropy-units(torch) array
...
```

Marray currently passes almost the entirety of https://github.com/data-apis/array-api-tests. It would be cool to see if these astropy-units namespaces could do the same!

This would bring a few benefits:
- for users, they can use astropy.units with whatever standard-compatible library they like
- for libraries which depend on astropy.units, they can also then work on supporting any standard-compatible arrays
- users can pass `Quantity`s to any library which supports standard-compatible arrays, like SciPy (eventually!)

**Comment by neutrinoceros** [comment]
Hi @lucascolley, we definitely agree that this is sensible, useful, and worth doing ! There's actually an undergoing effort, lead by @nstarman and @mhvk, to outsource `astropy.units` to an isolated package, and supporting the Array API is one of the primary goals of this initiative !

See https://github.com/astropy/astropy-APEs/pull/91 for the initial proposal and discussion, and https://github.com/astropy/quantity-2.0/ for the actual implementation (not ready for production at the time of writing !).

**Comment by lucascolley** [comment]
cool! @nstarman do you envision quantity-2.0 as covering the entire scope of Pint? I started a POC of this for Pint Quantities.

**Comment by mhvk** [comment]
@lucascolley - that is super nice to hear, also for our own [`Masked`](https://docs.astropy.org/en/latest/utils/masked/index.html) class.

If you are interested in joining our `Quantity 2.0` effort, that would be very welcome -- I've been meaning to push up coverage of the elementwise functions, maybe this is a good incentive to do it even if I feel there are a few loose ends.

**Comment by lucascolley** [comment]
Happy to join the effort @mhvk! (just can't guarantee any time commitments)

I'll let you know if I get stuck with anything on the Pint side. If any blockers arise then hopefully they can be solved for quantity-2.0

**Comment by pllim** [comment]
Hard to keep track... Is this also related to https://github.com/astropy/astropy/issues/13460 ? Thanks, all!

**Comment by mhvk** [comment]
> Hard to keep track... Is this also related to #13460 ? Thanks, all!

No, not directly - #13460 is an independent idea to make arrays themselves be quantity-like, by having the unit be part of the `dtype`, while here we are talking more about a real `Quantity` class interacting properly with different types of arrays.

I think the dtype idea is still worth pursuing, but it may be quite a bit harder, certainly as long as it remains impossible to define dtype in python (which would not be the end product, but so much easier to experiment with...).

**Comment by lucascolley** [comment]
closing this out as work is now ongoing over at @quantity-dev
